### PR TITLE
Make JoinError Sync

### PR DESCRIPTION
**Pull request by vorot93**
This is a fix for #1883.

Box and Mutex add overhead but they work and since it's not happy path, I figured they should suffice.

### CODE REVIEW

File: tokio/src/task/error.rs [hunk 1]
```diff
@@ -23,7 +23,7 @@ impl JoinError {
     }
 
     /// Create a new `panic` error
-    pub fn panic(err: Box<dyn Any + Send + 'static>) -> JoinError {
+    pub fn panic(err: Box<dyn Any + Send + Sync + 'static>) -> JoinError {
```
Comment (seanmonstar) [inline tokio/src/task/error.rs]: This is technically a breaking change, right?

Comment (vorot93) [reply tokio/src/task/error.rs]: I *could* put the mutex inside `Repr` and leave the interface unchanged, I guess

Comment (carllerche) [reply tokio/src/task/error.rs]: It could be deprecated and we add panic2?

Comment (seanmonstar) [reply tokio/src/task/error.rs]: Are public constructors even needed?

Comment (carllerche) [reply tokio/src/task/error.rs]: We could remove for now. I mostly added for completeness

Comment (vorot93) [reply tokio/src/task/error.rs]: I've restricted their to visibility the crate.

**Comment by jebrosen** [comment]
I might be missing something, but is there any way to even access the panic error? At first glance it looks like this switch from a `Box` that is never read to a `Mutex` that is never read.

**Comment by carllerche** [comment]
I imagine we will want to expose it in order to re-throw

**Pull request merged**